contracts/psm/VaultManager.sol

```diff
@@ -0,0 +1,229 @@
+        require(_psm != address(0), "psm cannot be zero address");
+        require(_token != address(0), "token cannot be zero address");
+        __AccessControl_init();
+        __UUPSUpgradeable_init();
+
+        _setupRole(DEFAULT_ADMIN_ROLE, _admin);
+        _setupRole(MANAGER, _manager);
+
+        psm = _psm;
+        token = _token;
+
+        emit SetPSM(_psm);
+        emit SetToken(_token);
+    }
+
+    modifier onlyPSM() {
+        require(msg.sender == psm, "Only PSM can call this function");
+        _;
+    }
+
+    /**
+      * @dev deposit token to adapters, only PSM can call this function
+      * @param amount deposit amount
+      */
+    function deposit(uint256 amount) external onlyPSM {
+        require(amount > 0, "deposit amount cannot be zero");
+
+        // transfer token to this contract
+        IERC20(token).safeTransferFrom(msg.sender, address(this), amount);
+        _distribute(amount);
+
+        emit Deposit(amount);
+    }
+
+    function _distribute(uint256 amount) private {
+        uint256 remain = amount;
+        uint256 totalPoint = getTotalPoint();
+
+        // deposit token to adapters by adapter point
+        for (uint256 i = 0; i < adapters.length; i++) {
+            if (adapters[i].active) { // only active adapter can be used
+                //  adapterAmount = depositAmount * point / totalPoint
+                uint256 amt = Math.mulDiv(amount, adapters[i].point, totalPoint);
+                if (amt > 0) {
+                    IERC20(token).safeIncreaseAllowance(adapters[i].adapter, amt);
+                    IAdapter(adapters[i].adapter).deposit(amt);
+                    remain -= amt;
+                }
+            }
+        }
+
+        // if remain amount > 0, add to localtoken
+        if (remain > 0) {
+            localToken += remain;
+        }
+    }
+
+    function getTotalPoint() public view returns (uint256) {
+        uint256 totalPoint;
+        for (uint256 i = 0; i < adapters.length; i++) {
+            if (adapters[i].active) {
+                totalPoint += adapters[i].point;
+            }
+        }
+
+        return totalPoint;
+    }
+
+    /**
+      * @dev withdraw token from adapters, only PSM can call this function
+      * @param receiver receiver address
+      * @param amount withdraw amount
+      */
+    function withdraw(address receiver, uint256 amount) external onlyPSM {
+        require(amount > 0, "withdraw amount cannot be zero");
+
+        uint256 remain = amount;
+
+        // withdraw token from local first
+        if (localToken >= remain) {
+            IERC20(token).safeTransfer(receiver, remain);
+            localToken -= remain;
+            return;
+        } else {
+            IERC20(token).safeTransfer(receiver, localToken);
+            remain -= localToken;
+            localToken = 0;
+        }
+
+        // withdraw token from adapters
+        for (uint256 i = 0; i < adapters.length; i++) {
```

Comment:
is it better to withdraw multiple source to the VaultMgr first, and then send to user as one shot?

Reply:
This will reduce the gas of a transfer